### DISCUSSION TITLE
Different Python versions

### DISCUSSION THREAD

**Discussion by Lewiscowles1986** [pyodide/pyodide#4311] [Q&A] [answered]
Thanks for making this amazing bit of software. Might there perhaps be support now or in the future for running different python versions under pyodide?

Lots of online services offer a scratch-pad to run python in the browser. Some use heavyweight software, such as guacamole, some forward to containers, but with pyodide, there is an opportunity to be able to make a version which could work offline.
I'm not thinking large projects of code here, just enough for learners to use in a browser.

I've just added to the StackOverflow answers on https://stackoverflow.com/questions/64669355/how-to-copy-download-file-created-in-pyodide-in-browser/64670850 for how to download a file. The last part I think I'd need to do, is be able to load a specific python version (Ideally 3.5 - 3.12) for a learner / journaller to use. I'm assuming this would be changing the url to pyodide; but I've not been able to find any instructions for what is involved in getting a specific version of python running.

**Comment by ryanking13** [comment, marked as answer]
Maybe in the future, yes, but I don't think it will happen soon.

The two main reasons are that

1) Pyodide is currently a project managed by a very small number of maintainers, so we simply don't have the resources, and

2) WebAssembly support in Python is still in a very beta stage and evolving very fast. For example, CPython just started to support WASM starting from 3.11, and lots of things will change in 3.12, 3.13, and so on. Also, the WASM compiler toolchain (Emscripten) is also evolving very fast, so a same Python version compiled with a different Emscripten may behave differently.

Therefore, it is almost impossible to maintain and manage multiple versions at the moment.

**Reply by EwoutH** [reply]
Do you plan at updating to Python 3.13 at some point? If so, how do you envision the timeline, roughly?

**Reply by Lewiscowles1986** [reply]
I don't know about pyodide, but it looks like the rustpython project (also can run in a browser via Wasm) might be another way to gt there.

**Reply by hoodmane** [reply]
We're currently on Python 3.12 we should update to 3.13 around May of 2025.

**Reply by ms-pablo-ruiz** [reply]
@hoodmane was python updated to 3.13? i see a couple of merges here and there, but i struggle to find any official announcement on this, actually I  find hard to find any statement about python versions in general both in the docs or the github. If anyone has the link with this info I would be grateful, thanks in advance

**Reply by hoodmane** [reply]
Main branch is 3.13 right now.

**Reply by ms-pablo-ruiz** [reply]
ah cool, thanks , do you think we should include that info in the docs/ readme or anywhere else? I don't mind to contribute myself to make that happen. I think is important to make it clear for totally new people like me.

**Reply by agriyakhetarpal** [reply]
Hi @ms-pablo-ruiz, the update is mentioned here: https://pyodide.org/en/latest/project/changelog.html#unreleased. It is a small entry at the moment; I think it could be made more prominent. CPython updates could have their own sections, I think. A PR to do this would be welcome. Thanks!

**Reply by ms-pablo-ruiz** [reply]
Yes @agriyakhetarpal, thanks for pointing that out, for me this looks like a major feature to brag about, not a minor deal, for instance having my code in python 3.13, is a great deal that now it will be supported. But for what i understand is still unreleased so we may as well wait a bit or simply include in the docs 3.12 for now.

About where to document the version it i can suggest:
- getting started.  https://pyodide.org/en/stable/usage/quickstart.html
- Downloading and deploying pyodide.
- pyodide python compatibility page: https://pyodide.org/en/stable/usage/wasm-constraints.html
And also:
- readme.md ,of this repo
- Announcement in the discussions tab once is realased
- Annoucement on discord? (not sure how active it is)

I understand that the pyodide version don't follow python ones being unrelated projects, but on the other hand it would be intuitive if they do follow the same numeration in some way or maybe there is a table where the pyodide version is matched to the python version it follows. let me know if this makes sense or not, I don't know yet if pyodide is fully python version agnostic.

What are your opinion should I proceed creating a PR changing all of these? I guess the official announcement should still be in your side.

Regards.

**Reply by hoodmane** [reply]
Yeah we're getting close to making a release. If you wanted to add a table to the readme showing which Python versions corerspond to which Pyodide version ranges that would be very helpful.

Obviously the announcements will need to come from us, but if you have opinions the draft of the release notes for Pyodide 0.28 are here, any comments you have are welcome.
https://github.com/pyodide/pyodide-blog/pull/53

**Reply by agriyakhetarpal** [reply]
I agree with @hoodmane. IMO, It would be nice to document everything in one place and link to it in other pages. It could be a separate page or perhaps be added to https://pyodide.org/en/stable/usage/wasm-constraints.html (the page title is "Pyodide Python compatibility", so maybe we could rename the file? Or would that introduce a lot of broken URLs when navigating between versions?). Announcing in the Discussions tab also sounds like a good idea. A table with historical compatibility notes is also super useful. I would not call Pyodide an unrelated project to Python (CPython), as there is shared code between the two repositories. It is not agnostic and does depend on one CPython version at a time (down to the micro version).

Re: Discord, we actively monitor it (enough) to respond to usage and development questions when possible, but I would suggest not announcing it there, as, in my experience, users are not keen on receiving `@everyone` or `@here` pings – they're quite intrusive.

**Reply by ms-pablo-ruiz** [reply]
lgtm.

Is there already a easy way to check the versions for making such a table? or do we start the table with this version and forget about the old ones?

Do you think that new pyiodide versions can be paired to the Cpython ones? Like if we now support 3.13.1, we create pyodide 3.13.1 and maybe if we need to change something 3.13.1post2 (the post keyword is the only way i know pep compliant of having an additional number) though this doesn't look much like semver... probably is a bad idea.

**Reply by agriyakhetarpal** [reply]
> Is there already a easy way to check the versions for making such a table? or do we start the table with this version and forget about the old ones?

Yes, you can use https://github.com/pyodide/pyodide/blob/074948ffca1208e5dc249e821b4f398b383b49a9/Makefile.envs#L1 to check the Python version across various tags in the repository. Perhaps we can list the Pyodide versions corresponding to the [non-EOL versions for CPython](https://endoflife.date/python), so that would be Python 3.9 and later.

> Do you think that new pyiodide versions can be paired to the Cpython ones? Like if we now support 3.13.1, we create pyodide 3.13.1 and maybe if we need to change something 3.13.1post2 (the post keyword is the only way i know pep compliant of having an additional number) though this doesn't look much like semver... probably is a bad idea.

It is a possibility indeed (Cython also released 0.29.x and then switched to 3.x), but I wouldn't be keen on this. My reason is that [PEP 2026](https://peps.python.org/pep-2026/) (slightly different but broadly similar in concept, centred more around calendar versioning) was similarly rejected on the grounds that it would create churn by breaking the sequential and monotonic nature of software versioning. If you have a refined proposal that could cover all edge cases, please feel free to open a new issue about it – it can be explored further, definitely. Thanks!

**Reply by hoodmane** [reply]
There has been some discussion about maintaining an overlap period of 6 months where we have a versions of Pyodide using two different Python versions. That might require a change to our versioning scheme.

**Reply by hoodmane** [reply]
There's a table here of the PRs where we made the version updates:
https://pyodide.org/en/latest/development/maintainers.html#old-major-python-upgrades

**Reply by Lewiscowles1986** [reply]
> I would suggest not announcing it there, as, in my experience, users are not keen on receiving &#64;everyone or &#64;here pings – they're quite intrusive.

They are, and for older people, who use discord, but despise it (me), it is very off-putting when folks bulk message.
